File: pakalon-cli/src/providers/metrics.ts
/**
 * Metrics Collection for Pakalon CLI
 *
 * Tracks usage metrics for monitoring and analytics.
 */

import logger from "@/utils/logger.js";
import { isFeatureEnabled } from "@/config/features.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface Metric {
  name: string;
  value: number;
  labels: Record<string, string>;
  timestamp: Date;
}

export interface MetricConfig {
  enabled: boolean;
  endpoint?: string;
  apiKey?: string;
  batchSize?: number;
  flushIntervalMs?: number;
}

export interface ProviderMetric {
  providerId: string;
  latencyMs: number;
  success: boolean;
  tokens?: number;
  model?: string;
}

export interface TokenMetric {
  providerId: string;
  modelId: string;
  inputTokens: number;
  outputTokens: number;
}

// ---------------------------------------------------------------------------
// Default Configurations
// ---------------------------------------------------------------------------

const DEFAULT_CONFIG: MetricConfig = {
  enabled: false,
  batchSize: 10,
  flushIntervalMs: 30000,
};

// ---------------------------------------------------------------------------
// Metrics Collector
// ---------------------------------------------------------------------------

export class MetricsCollector {
  private metrics: Metric[] = [];
  private config: MetricConfig;
  private flushTimer: ReturnType<typeof setInterval> | null = null;

  constructor(config: Partial<MetricConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };

    // Check if analytics feature is enabled
    if (isFeatureEnabled("analytics")) {
      this.config.enabled = true;
    }

    if (this.config.enabled && this.config.flushIntervalMs) {
      this.startAutoFlush();
    }
  }

  /**
   * Record a metric
   */
  recordMetric(name: string, value: number, labels: Record<string, string> = {}): void {
    if (!this.config.enabled) {
      return;
    }

    const metric: Metric = {
      name,
      value,
      labels,
      timestamp: new Date(),
    };

    this.metrics.push(metric);

    if (this.metrics.length >= (this.config.batchSize || 10)) {
      this.flush();
    }
  }

  /**
   * Record provider latency
   */
  recordProviderLatency(providerId: string, latencyMs: number, success: boolean): void {
    this.recordMetric("provider_latency_ms", latencyMs, {
      provider: providerId,
      success: success.toString(),
    });
  }

  /**
   * Record token usage
   */
  recordTokenUsage(providerId: string, modelId: string, inputTokens: number, outputTokens: number): void {
    this.recordMetric("token_usage", inputTokens + outputTokens, {
      provider: providerId,
      model: modelId,
      type: "total",
    });

    this.recordMetric("token_usage", inputTokens, {
      provider: providerId,
      model: modelId,
      type: "input",
    });

    this.recordMetric("token_usage", outputTokens, {
      provider: providerId,
      model: modelId,
      type: "output",
    });
  }

  /**
   * Record provider health
   */
  recordProviderHealth(providerId: string, healthy: boolean, latencyMs: number): void {
    this.recordMetric("provider_health", healthy ? 1 : 0, {
      provider: providerId,
    });

    this.recordMetric("provider_latency_ms", latencyMs, {
      provider: providerId,
    });
  }

  /**
   * Record error
   */
  recordError(providerId: string, errorType: string, errorMessage: string): void {
    this.recordMetric("error_count", 1, {
      provider: providerId,
      error_type: errorType,
      error_message: errorMessage,
    });
  }

  /**
   * Record feature usage
   */
  recordFeatureUsage(feature: string, usage: number = 1): void {
    this.recordMetric("feature_usage", usage, {
      feature,
    });
  }

  /**
   * Start auto-flush timer
   */
  private startAutoFlush(): void {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
    }

    this.flushTimer = setInterval(() => {
      if (this.metrics.length > 0) {
        this.flush();
      }
    }, this.config.flushIntervalMs);
  }

  /**
   * Stop auto-flush timer
   */
  stopAutoFlush(): void {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }
  }

  /**
   * Flush metrics to endpoint
   */
  async flush(): Promise<void> {
    if (this.metrics.length === 0) {
      return;
    }

    const metricsToSend = [...this.metrics];
    this.metrics = [];

    if (!this.config.endpoint) {
      // Log metrics instead of sending
      logger.debug(`[Metrics] Flushing ${metricsToSend.length} metrics`);
      for (const metric of metricsToSend) {
        logger.debug(`[Metrics] ${metric.name}: ${metric.value}`, metric.labels);
      }
      return;
    }

    try {
      await fetch(this.config.endpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...(this.config.apiKey ? { Authorization: `Bearer ${this.config.apiKey}` } : {}),
        },
        body: JSON.stringify({ metrics: metricsToSend }),
      });
    } catch (error) {
      logger.error(`[Metrics] Failed to flush metrics: ${error}`);
      // Re-add failed metrics
      this.metrics.unshift(...metricsToSend);
    }
  }

  /**
   * Get metrics count
   */
  getMetricsCount(): number {
    return this.metrics.length;
  }

  /**
   * Clear metrics
   */
  clear(): void {
    this.metrics = [];
  }
}

// Global instance
export const metricsCollector = new MetricsCollector();

// ---------------------------------------------------------------------------
// Convenience Functions
// ---------------------------------------------------------------------------

/**
 * Record a metric
 */
export function recordMetric(name: string, value: number, labels: Record<string, string> = {}): void {
  metricsCollector.recordMetric(name, value, labels);
}

/**
 * Record provider latency
 */
export function recordProviderLatency(providerId: string, latencyMs: number, success: boolean): void {
  metricsCollector.recordProviderLatency(providerId, latencyMs, success);
}

/**
 * Record token usage
 */
export function recordTokenUsage(
  providerId: string,
  modelId: string,
  inputTokens: number,
  outputTokens: number
): void {
  metricsCollector.recordTokenUsage(providerId, modelId, inputTokens, outputTokens);
}

/**
 * Record provider health
 */
export function recordProviderHealth(providerId: string, healthy: boolean, latencyMs: number): void {
  metricsCollector.recordProviderHealth(providerId, healthy, latencyMs);
}

/**
 * Record error
 */
export function recordError(providerId: string, errorType: string, errorMessage: string): void {
  metricsCollector.recordError(providerId, errorType, errorMessage);
}

/**
 * Record feature usage
 */
export function recordFeatureUsage(feature: string, usage: number = 1): void {
  metricsCollector.recordFeatureUsage(feature, usage);
}

/**
 * Flush metrics
 */
export async function flushMetrics(): Promise<void> {
  return metricsCollector.flush();
}

File: pakalon-backend/app/metrics.py
"""Metrics collection for Pakalon backend.

Tracks usage metrics for monitoring and analytics.
"""

import logging
import time
from datetime import datetime
from typing import Any

from pydantic import BaseModel

from app.config import get_settings
from app.features import get_feature_flags

logger = logging.getLogger(__name__)


class Metric(BaseModel):
    """A single metric data point."""

    name: str
    value: float
    labels: dict[str, str] = {}
    timestamp: datetime


class ProviderMetric(BaseModel):
    """Provider-specific metric."""

    provider_id: str
    latency_ms: float
    success: bool
    tokens: int | None = None
    model: str | None = None


class TokenMetric(BaseModel):
    """Token usage metric."""

    provider_id: str
    model_id: str
    input_tokens: int
    output_tokens: int


class MetricsCollector:
    """Collects and stores metrics for monitoring."""

    def __init__(self, max_buffer_size: int = 100):
        self._metrics: list[Metric] = []
        self._provider_metrics: list[ProviderMetric] = []
        self._token_metrics: list[TokenMetric] = []
        self._max_buffer_size = max_buffer_size

    def record_metric(self, name: str, value: float, labels: dict[str, str] | None = None) -> None:
        """Record a metric."""
        metric = Metric(
            name=name,
            value=value,
            labels=labels or {},
            timestamp=datetime.now(),
        )
        self._metrics.append(metric)

        if len(self._metrics) >= self._max_buffer_size:
            self._flush_metrics()

    def record_provider_latency(self, provider_id: str, latency_ms: float, success: bool) -> None:
        """Record provider latency."""
        metric = ProviderMetric(
            provider_id=provider_id,
            latency_ms=latency_ms,
            success=success,
        )
        self._provider_metrics.append(metric)

        # Also record as generic metric
        self.record_metric(
            "provider_latency_ms",
            latency_ms,
            {"provider": provider_id, "success": str(success)},
        )

    def record_token_usage(self, provider_id: str, model_id: str, input_tokens: int, output_tokens: int) -> None:
        """Record token usage."""
        metric = TokenMetric(
            provider_id=provider_id,
            model_id=model_id,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
        )
        self._token_metrics.append(metric)

        total = input_tokens + output_tokens
        self.record_metric(
            "token_usage",
            total,
            {"provider": provider_id, "model": model_id},
        )

    def record_provider_health(self, provider_id: str, healthy: bool, latency_ms: float) -> None:
        """Record provider health status."""
        self.record_metric(
            "provider_health",
            1.0 if healthy else 0.0,
            {"provider": provider_id},
        )

    def record_error(self, provider_id: str, error_type: str, error_message: str) -> None:
        """Record an error."""
        self.record_metric(
            "error_count",
            1.0,
            {"provider": provider_id, "error_type": error_type},
        )

    def _flush_metrics(self) -> None:
        """Flush metrics buffer (log or send to external service)."""
        if not self._metrics:
            return

        settings = get_settings()

        # In development, just log metrics
        if settings.is_development:
            for metric in self._metrics[-5:]:  # Log last 5 metrics
                logger.debug(f"Metric: {metric.name}={metric.value} {metric.labels}")

        # Clear buffer
        self._metrics.clear()

    def get_stats(self) -> dict[str, Any]:
        """Get collector statistics."""
        return {
            "buffered_metrics": len(self._metrics),
            "total_provider_metrics": len(self._provider_metrics),
            "total_token_metrics": len(self._token_metrics),
        }


# Global instance
metrics_collector = MetricsCollector()
